src/roblox_api/cframe.rs

```diff
@@ -0,0 +1,112 @@
+use rlua::{UserData, UserDataMethods, Value as LuaValue};
+
+use crate::value::{CFrameValue, Vector3Value};
+
+pub struct CFrame;
+
+impl CFrame {
+    fn from_position(x: f32, y: f32, z: f32) -> CFrameValue {
+        CFrameValue::new(rbx_dom_weak::types::CFrame::new(
+            rbx_dom_weak::types::Vector3::new(x as f32, y as f32, z as f32),
+            // TODO: replace with `rbx_dom_weak::types::Matrix3::identity()` once
+            // a version higher than 0.3.0 of rbx_types ships
+            rbx_dom_weak::types::Matrix3::new(
+                rbx_dom_weak::types::Vector3::new(1.0, 0.0, 0.0),
+                rbx_dom_weak::types::Vector3::new(0.0, 1.0, 0.0),
+                rbx_dom_weak::types::Vector3::new(0.0, 0.0, 1.0),
+            ),
+        ))
+    }
+}
+
+impl UserData for CFrame {
+    fn add_methods<'lua, T: UserDataMethods<'lua, Self>>(methods: &mut T) {
+        methods.add_function(
+            "new",
+            |_context,
+             triplet: (
+                Option<LuaValue<'_>>,
+                Option<LuaValue<'_>>,
+                Option<LuaValue<'_>>,
+            )| {
+                match triplet {
+                    (None, None, None) => Ok(Self::from_position(0.0, 0.0, 0.0)),
+                    (Some(LuaValue::Number(x)), None, None) => {
```

Comment:
There's gotta be a better way than all this, especially when you'll probably want ALL the CFrame constructors.

Reply:
I just pushed something that makes it better, but let me know if you have something else in mind! Not sure what to call the `Either` type so if you have a better name let me know 🙂 

Thanks for your help by the way 👍

Reply:
I think we can simplify this code a lot using option combinators. I threw together some rough code that probably doesn't compile, but try something along these lines:

```rust
fn try_into_f32(value: LuaValue<'_>) -> Option<f32> {
    match value {
        LuaValue::Number(num) => Some(num as f32),
        LuaValue::Integer(int) => Some(int as f32),
        _ => None,
    }
}

fn to_position(
    triplet: (
        Option<LuaValue<'_>>,
        Option<LuaValue<'_>>,
        Option<LuaValue<'_>>,
    ),
) {
    match &triplet {
        (Some(LuaValue::UserData(user_data)), None, None) => {
            let position = &*user_data.borrow::<Vector3Value>()?;

            return Ok(CFrameValue::new(CFrame::new(
                position.0,
                // TODO: replace with `rbx_dom_weak::types::Matrix3::identity()` once
                // a version higher than 0.3.0 of rbx_types ships
                Matrix3::new(
                    Vector3::new(1.0, 0.0, 0.0),
                    Vector3::new(0.0, 1.0, 0.0),
                    Vector3::new(0.0, 0.0, 1.0),
                ),
            )));
        }

        _ => {}
    }

    let x = triplet.0.and_then(try_into_f32);
    let y = triplet.1.and_then(try_into_f32);
    let z = triplet.2.and_then(try_into_f32);

    match (x, y, z) {
        (Some(x), Some(y), Some(z)) => Ok(Self::from_position(x, y, z)),
        _ => Err(rlua::Error::external(
            "invalid argument #1 to 'new' (Vector3 expected)",
        )),
    }
}
```